feat: epd-visualization integration library

sam-garland opened the pull request:
Changed branch name as requested: epic/epd-visualization-> integration-libs/epd-visualization

Previous PR before branch change: #14423

sam-garland committed:
feat: epd-visualization integration library (#13933)

sam-garland committed:
chore: merge from develop (#14473)

* feat: display done button in configurator displayOnly view (14326)

The button was previously disabled. As of 5.0, it's present, showing 'done', and navigates back to order history or checkout review order page 
closes #14152

* test: fix flaky configurator e2e test (#14413)

Needs adaption to a11y changes
closes #14412

* fix: improve checkboxes for new payment during checkout (#14032)

* fix an issue in express checkout (#14418)

* test: Enhance test coverage for textfield configurator (#14403)

Enhance test coverage for textfield configurator regarding HTML extension related to a11y / screen reader improvements

Closes #14402

* test: re-enable e2e vc configurator cart test (#14408)

Test was marked flaky. Since we transitioned to a new 21.05 based backend, it can be activated again

Closes #12274

* chore: Schematics migration 5.0 scaffold (#14411)

Closes: #14406

* fix: Fix pdp heading levels (#13484)

* test: Test is still flaky

* test: ASM e2e flakiness

* chore: provide 5.0 migration schematics for addToCart component (#14451)

closes #14379

* test: fix CPQ configurator order process test (#14410)

Closes #14389

* fix: I18n back-end config (#14453)

closes #14458

Co-authored-by: Kornkanok Changsuwan (Sia) <79838431+Changsuwan@users.noreply.github.com>
Co-authored-by: Christoph Hinssen <33626130+ChristophHi@users.noreply.github.com>
Co-authored-by: Benjamin Emiliani <49131189+BenjaminEmiliani@users.noreply.github.com>
Co-authored-by: Weizheng Gao <44440575+WeizhengSap@users.noreply.github.com>
Co-authored-by: LarisaStar <61147963+Larisa-Staroverova@users.noreply.github.com>
Co-authored-by: Patryk Smul <psmul@divante.pl>
Co-authored-by: Michał D <mdydo@divante.pl>
Co-authored-by: Gilberto Alvarado <gilberto.alvarado.jimenez@sap.com>
Co-authored-by: MichaelBoeser <54353128+MichaelBoeser@users.noreply.github.com>
Co-authored-by: Nikola Zarić <nikola.zaric.ns@gmail.com>

sam-garland committed:
chore: cherry-pick from epic/epd-visualization-maintenance (#62) (#14476)

* feat: epd-visualization integration library (#13933)

* chore: cherry-pick from epic/epd-visualization-maintenance

* Moved styles and fixed lazy loading for schematic updates

* chore: moved styles into epd-visualization library

* fix: lazy loading wasn't working in schematic

* fix: added missing peer dependencies.

* chore: switched to a floating UI5 version

* chore: change schematic propmt/default.
Don't use ACC environment as default.

* chore: ran prettier

* fix: synchronise selection made before scene loaded

sam-garland committed:
chore "chore: merge from develop (#14473)" (#14477)

This reverts commit 6b3f16b9d006257968acb7f7c418cd12c35a6983.

sam-garland committed:
Merge branch 'develop' into epic/epd-visualization

sam-garland committed:
chore: cherry pick from epic/epd-visualization-maintenance (#14516)

* feat: epd-visualization integration library

* chore: cherry-pick from epic/epd-visualization-maintenance

* Moved styles and fixed lazy loading for schematic updates

* chore: moved styles into epd-visualization library

* fix: lazy loading wasn't working in schematic

* fix: added missing peer dependencies.

* chore: switched to a floating UI5 version

* chore: change schematic propmt/default.
Don't use ACC environment as default.

* chore: ran prettier

* fix: synchronise selection made before scene loaded

* chore: cherry pick from epic/epd-visualization-maintenance

* chore: use powertools-epdvisualization-spa site instead of powertools-spa.

* Updated README.md

* fix: update e2e tests for new site

* fix: update UI when product references change

* fix: update site for e2e tests

* fix: animation slider visible in 2D mode.

sam-garland committed:
chore: allow vscode to resolve package imports (#14518) (#14520)

* chore: allow vscode to resolve package imports

* tsconfig.json & tsconfig.eslint.json not needed

sam-garland committed:
Only use products from sample data in e2e tests

sam-garland committed:
chore: allow vscode to resolve package imports

sam-garland committed:
chore: removed currently unused type definitions (#14524)

sam-garland committed:
chore: remove REST services from public API. (#14526)

sam-garland committed:
fix: hotspot selection colour was opaque (#14528)

sam-garland committed:
fix: show spinner until geometry streamed in (#14530)

sam-garland committed:
fix: go to list start when filtered set changes (#14532)

This fixes a bug in which results are not always shown when
filtering performed when on a page other than the first page
in the spare parts list.

sam-garland committed:
Merge branch 'develop' into epic/epd-visualization

sam-garland committed:
chore: specify thumbnail format for thumbnails

sam-garland committed:
fix: filter unresolvable nodes (#14555)

sam-garland committed:
chore: use UI5 1.97.0 (#14552)

sam-garland committed:
Merge branch 'develop' into epic/epd-visualization

sam-garland committed:
chore: changes request in PR review (#14567)

* chore: move translations onto own namespace
Use a single file for translations for the epd-visualization integration library

* chore: removed start:local yarn script

* chore: changed component selectors
To avoid any collision with components outside integration library.

* chore: added inner object to configuration

This is to avoid conflicts in the global namespace for configuration.

* chore: added config validators

* chore: do not include getTestConfig in public API

* chore: ran prettier

* chore: update prefix

sam-garland committed:
fix: do not modify opacity on reference nodes. (#14570)

We only want to modify nodes that are part of the normal scene hierachy.
Jira: https://sapjira.wdf.sap.corp/browse/EPDVISUALIZATION-3194

sam-garland committed:
chore: moved code from component to service (#14572)

sam-garland committed:
fix: cleanup subscriptions (#14583)

sam-garland committed:
fix: unsubscribe on load failure (#14584)

sam-garland committed:
fix: fixed issues affecting product page navigation (#14586)

* fix: ensure maps updated on scene change

* fix: use shared Observable for getProductReferences$()

This avoids redundant API calls being made.
Removed shareReplay() from currentProduct$. This was causing issues when navigating between product pages.

* chore: ran prettier

sam-garland committed:
chore: review changes (#14592)

* chore: renamed highlighted -> selected

* chore: remove @Input, @Output from service code

* chore: removed unnecessary allowUmdGlobalAccess configuration.

* chore: added jsdoc to configuration types

* fix: keyboard navigation requirements

* chore: removed pointless code

* chore: ran prettier

sam-garland committed:
fix: timing issue broke selection in viewport (#14594)

sam-garland committed:
chore: ran organize imports (#14596)

sam-garland committed:
chore: review changes (#14597)

* chore: used destructuring

* chore: use optional chaining

* chore: renaming

sam-garland committed:
chore: added parameter checking to test (#14603)

sam-garland committed:
chore: use adapter/connector model for REST APIs (#14601)

sam-garland committed:
chore: added default config (#14610)

sam-garland committed:
chore: review changes (#14616)

sam-garland committed:
chore: remove need for skipLibCheck in customer app (#14629)

* chore: got rid of implicit any warning.

* chore: remove need for skipLibCheck in customer app

* chore: ignore "unused" @sapui5/ts-types-esm lib

* chore: add visual-viewer.service.ts to public api

* chore: fix implicit any warnings

sam-garland committed:
fix: add missing code to install script (#14633)

sam-garland committed:
chore: add sample data assets (#14632)

sam-garland committed:
fix: fixed spelling in script (#14635)

* chore: added missing return types

* fix: fix spelling issue

* chore: made focus outline look nice on list items

sam-garland committed:
fix: fix spelling issue (#14636)

* chore: added missing return types

* fix: fix spelling issue

* chore: made focus outline look nice on list items

sam-garland committed:
chore: implicit any warnings (#14638)

sam-garland committed:
fix: hotspot filtering when hotspot nodes have descendant hotspot nodes (#14646)

sam-garland committed:
fix: install script fixes (#14656)

* fix: fix default feature list

* fix: added code to set base site

sam-garland committed:
chore: revert change to remove implicit any warnings (#14659)

This reverts commit f4cd2e1ad1024e7fcd29a97f3de5012bc76bce19.

sam-garland committed:
chore: specify metadata source in lookup. (#14661)

Using new syntax that is available in the 2111 release:
https://github.wdf.sap.corp/vis/storage-service/blob/e289d322c7ad34d72e0020cb6e76827b21a783b6/doc/VISTOR_Public.yaml#L7133

sam-garland committed:
Merge branch 'develop' into epic/epd-visualization

sam-garland committed:
chore: remove redundant configuration. (#14663)

sam-garland committed:
Merge branch 'develop' into merge-from-develop (#14664)

sam-garland committed:
chore: removed redundant deletions/unwanted changes (#14666)

* chore: removed redundant deletions

* chore: reverted changes unrelated to PR

* chore: updated a comment

* chore: added comments

sam-garland committed:
chore: ran prettier (#14667)

sam-garland committed:
fix: avoid performance bug when restoring opacity (#14671)

sam-garland committed:
chore: merge from develop (#14674)

* test: Add tabScreenshot() method for a11y tab e2e testing (#13818)

Closes: #13686

* feat: disable shipping method radio input when there is ongoing delivery mode request (#14401)

* add spinner when there is onging delivery mode request

* disable radio input instead of using spinner

* fix regression error: only retrieve enabled radio input

* fix broken a18y test

* when loading shipping methods, continue button hiden by spinner

* remove unused flag

* Michal's improvement

* fix broken e2e test

* fix prettier issue

* disable fieldset

* add feature level

* fix unit test after feature level is introduced

* small refactor

* remove log

* add doc for breaking change

* up

* fix disable delay issue

* remove console.log

* add unit test

* add space in comment

* update migration document

* fix build error

* remove duplicate spyOn code

* use BehaviorSubject to improve test

* emit default data in beforeEach

Co-authored-by: Michał Gruca <pucek9@gmail.com>

Co-authored-by: Caine Rotherham <rotherham.c@gmail.com>
Co-authored-by: Jerry Wang <58975336+wangzixi-diablo@users.noreply.github.com>
Co-authored-by: Michał Gruca <pucek9@gmail.com>

sam-garland committed:
Merge branch 'develop' into epic/epd-visualization

sam-garland committed:
fix: add guards for SSR execution path (#14688)

sam-garland committed:
chore: make e2e job conditional on branch

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization

sam-garland committed:
chore: add visualization data asset

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization

sam-garland committed:
chore: fix merge

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization

sam-garland committed:
[FIX] add B2B config in epd-visualization schematic (#14838)

sam-garland committed:
chore: match existing feature level

So unit tests pass.

sam-garland committed:
Revert "chore: match existing feature level"

This reverts commit 11041a13c3e76ddb333068991436e9e2e324bda0.

sam-garland committed:
Revert "[FIX] add B2B config in epd-visualization schematic (#14838)"

This reverts commit f931f850c9722520bd91449edf5b91ad0c2bfc21.

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization

Platonn commented:
Hi @sam-garland 
Would you mind resolving the conflicts in this PR do develop, please?
I'm going do the last review comparing the changes: `this-PR --> develop` vs. `merged-PR --> release/4.3.x` branch.
Thanks in advance

sam-garland committed:
chore: merge from develop (#14967)

* Bringing back codeowners

* feat: Introduce the cart base feature library (#13230)

(GH-13044)
The code for the base cart feature is extracted from the core Spartacus libraries and bundled in the `@spartacus/cart/base` entry point of the existing cart library.
By default, the base cart feature is lazy loaded for anonymous users.

* chore: Document breaking changes for A11Y for variant configuration (#14650)

 breaking A11Y changes for variant config

Closes #14611

* chore: Notification preferences e2e small refactor (#14904)

* chore: increase timeout on list and assignment to avoid flaky runs (#14865)

Closes #14347

* refactor: clean up components and increase test coverage (#14929)

Cleans up multi and single select bundle components.

Closes #14928

* fix: Add live region for status message (#14897)

Adds aria-live region for configuration update status message to be read by screen-readers. 

Closes #14862

* test: fix failing CPQ configurator e2e test (#14944)

Closes #14941

* fix: Fix for stock notification e2e test, merge two test in one with viewport context way (#14904)

* chore: add breaking changes of FacetList to 5.0md (#14934)

* fix: improve sanitization of product text in show more component (#14951)

Changes view in order to properly sanitize product text. Adds tests that validate sanitization.
Closes #14943

* Ran yarn:generate-deps

* Getting tests working.

* chore: update schematic tests

Co-authored-by: Gilberto Alvarado <gilberto.alvarado.jimenez@sap.com>
Co-authored-by: pla <plabadie@users.noreply.github.com>
Co-authored-by: steinsebastian <54309425+steinsebastian@users.noreply.github.com>
Co-authored-by: Michał D <mdydo@divante.pl>
Co-authored-by: Michal Szczepaniak <97286147+szczepaniak-michal@users.noreply.github.com>
Co-authored-by: Christoph Hinssen <33626130+ChristophHi@users.noreply.github.com>
Co-authored-by: Benjamin Emiliani <49131189+BenjaminEmiliani@users.noreply.github.com>

sam-garland committed:
chore: merge from develop (#14968)

* Bringing back codeowners

* feat: Introduce the cart base feature library (#13230)

(GH-13044)
The code for the base cart feature is extracted from the core Spartacus libraries and bundled in the `@spartacus/cart/base` entry point of the existing cart library.
By default, the base cart feature is lazy loaded for anonymous users.

* chore: Document breaking changes for A11Y for variant configuration (#14650)

 breaking A11Y changes for variant config

Closes #14611

* chore: Notification preferences e2e small refactor (#14904)

* chore: increase timeout on list and assignment to avoid flaky runs (#14865)

Closes #14347

* refactor: clean up components and increase test coverage (#14929)

Cleans up multi and single select bundle components.

Closes #14928

* fix: Add live region for status message (#14897)

Adds aria-live region for configuration update status message to be read by screen-readers. 

Closes #14862

* test: fix failing CPQ configurator e2e test (#14944)

Closes #14941

* fix: Fix for stock notification e2e test, merge two test in one with viewport context way (#14904)

* chore: add breaking changes of FacetList to 5.0md (#14934)

* fix: improve sanitization of product text in show more component (#14951)

Changes view in order to properly sanitize product text. Adds tests that validate sanitization.
Closes #14943

* chore: add codeowners for feature-libs and product-configurator/CPQ

Co-authored-by: Gilberto Alvarado <gilberto.alvarado.jimenez@sap.com>
Co-authored-by: pla <plabadie@users.noreply.github.com>
Co-authored-by: steinsebastian <54309425+steinsebastian@users.noreply.github.com>
Co-authored-by: Michał D <mdydo@divante.pl>
Co-authored-by: Michal Szczepaniak <97286147+szczepaniak-michal@users.noreply.github.com>
Co-authored-by: Christoph Hinssen <33626130+ChristophHi@users.noreply.github.com>
Co-authored-by: Benjamin Emiliani <49131189+BenjaminEmiliani@users.noreply.github.com>
Co-authored-by: Krzysztof Platis <platonn.git@gmail.com>

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization

sam-garland commented:
> Hi @sam-garland Would you mind resolving the conflicts in this PR do develop, please? I'm going do the last review comparing the changes: `this-PR --> develop` vs. `merged-PR --> release/4.3.x` branch. Thanks in advance

@Platonn I have changed the behaviour so that `cx-epd-visualization-compact-add-to-cart` buttons are not rendered when the `addToCart` feature is not configured (since these buttons will not do anything when clicked when the `addToCart` feature is not configured). 

I also notice that more manual steps are needed after the schematic runs.
After adding `provideConfig(defaultB2bOccConfig), provideConfig(defaultB2bCheckoutConfig)` to the `providers` section of `spartacus-configuration.module.ts` (and adding imports), I needed to add the `@spartacus/setup`, `@spartacus/checkout` and `@spartacus/order` packages to the `dependencies` section in the `package.json` and run `yarn install` before building.

sam-garland committed:
chore: UI5 version 1.97.0 -> 1.98.0 (#14969)

sam-garland committed:
Hide cx-epd-visualization-compact-add-to-cart when addToCart feature not configured (#14973)

Platonn commented:
@sam-garland 

> I have changed the behaviour so that cx-epd-visualization-compact-add-to-cart buttons are not rendered when the addToCart feature is not configured (since these buttons will not do anything when clicked when the addToCart feature is not configured).

Thank you for paing attention to the recent changes in develop with regards to the cart lib. That said, I'm not 100% sure the need to check if add-to-cart feature was installed in _this specific case._ Let me first consult with @plabadie and @WeizhengSap (authors of recent changes in the cart lib) and I'll get back to you.

Regarding the manual steps for installing b2b... understood, this is not comfortable situation for you. 

I would like to wait with deciding  whether we should add officialy an option `--b2b` to `@spartacus/schematics` installation schematics. It's becasue we plan in the future to recognize it automatically from the basesites configuration in the CMS.

Did you consider any other workarounds? For example:
a) writing your own script (kind of "internal schematics?") OR
b) change sample data: put your sample products into the electronics' product catalogue instead of the powertools' catalogue? then for testing purposes you could use the electronics' site OR
c) ...or maybe you have any other ideas for workaround?

sam-garland commented:
@Platonn 

> Thank you for paing attention to the recent changes in develop with regards to the cart lib. That said, I'm not 100% sure the need to check if add-to-cart feature was installed in _this specific case._ Let me first consult with @plabadie and @WeizhengSap (authors of recent changes in the cart lib) and I'll get back to you.

Sure.

> Regarding the manual steps for installing b2b... understood, this is not comfortable situation for you.
> 
> I would like to wait with deciding whether we should add officialy an option `--b2b` to `@spartacus/schematics` installation schematics. It's becasue we plan in the future to recognize it automatically from the basesites configuration in the CMS.
> 
> Did you consider any other workarounds? For example: a) writing your own script (kind of "internal schematics?") OR b) change sample data: put your sample products into the electronics' product catalogue instead of the powertools' catalogue? then for testing purposes you could use the electronics' site OR c) ...or maybe you have any other ideas for workaround?

Given that the long term plan is to detect the channel from the base site configuration, perhaps the `@spartacus/epd-visualization` schematic could:
- Read the `spartacus-configuration.module.ts` file to get the `baseSite` configuration
- Determine whether a B2B site is being used based on the name (defaulting to B2C)
- Add B2B configuration if required

Then the logic above could be replaced in the future if there is a more reliable way of determining whether the configured base site is a B2B site.

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization

Platonn commented:
Hi @sam-garland 

### Suggestion to remove some unnecessary code
Since you just extend the TS class `AddToCartComponent` and do nothing more around the add-to-cart feature, I believe we can get rid of some unnecessary code:
- in your component:
  - unnecessary check whether feature was configured
  - unnecessary custom constructor (with FeatureModulesService dependency):
  https://github.com/SAP/spartacus/blob/bc49b9dcaf38db53914a67f0088825ccebbb1d30/integration-libs/epd-visualization/components/visual-picking/visual-picking-tab/product-list/compact-add-to-cart/compact-add-to-cart.component.ts#L28-L51
  - ... and the check for `requiredFeaturesEnabled` in the template
  https://github.com/SAP/spartacus/blob/bc49b9dcaf38db53914a67f0088825ccebbb1d30/integration-libs/epd-visualization/components/visual-picking/visual-picking-tab/product-list/compact-add-to-cart/compact-add-to-cart.component.html#L2
... and in your modules you don't need to import the original `AddToCartModule` anytime:
  - https://github.com/SAP/spartacus/blob/34c9a5a4d78f76de931e89ef7c305d3b93c2fcc5/integration-libs/epd-visualization/components/visual-picking/visual-picking-tab/product-list/compact-add-to-cart/compact-add-to-cart.module.ts#L29
  - https://github.com/SAP/spartacus/blob/34c9a5a4d78f76de931e89ef7c305d3b93c2fcc5/integration-libs/epd-visualization/components/visual-picking/visual-picking-tab/product-list/visual-picking-product-list.module.ts#L25
  - https://github.com/SAP/spartacus/blob/34c9a5a4d78f76de931e89ef7c305d3b93c2fcc5/integration-libs/epd-visualization/components/visual-picking/visual-picking-tab/product-list/visual-picking-product-list.module.ts#L29

### Explanation:
1. No need to import `AddToCartModule`
   If you referenced the original component via the selector `<cx-add-to-cart>` in the html of some of your components, you would need to import the original `AddToCartModule`. But it's not the case for you. You just extend a TS class. 

2. No need to check if a feature was configured
   You don't relay in any way on the fact that the `ADD_TO_CART_FEATURE` is configured for lazy loading (in other words, you don't need to check if `featureModulesService.isConfigured(ADD_TO_CART_FEATURE)`).

    More info, how lazy loading works in your case:
    You just import a TS class `import { xxx } from '@spartacus/different-lib-that-might-be-lazy-loaded-separately`). Then the `xxx` class (which in your case is `AddToCartComponent`) will be extracted by Webpack during the compilation into a separate shared JS chunk. Then in runtime the Webpack will trigger loading of the shared chunk, when either the chunk of your EPD's feature module is lazy-loaded or JS chunks of other features importing the `AddToCartComponent` class are lazy-loaded (including the default module of ADD_TO_CART_FEATURE). 

### Note on a possible, completely different solution (which I don't want to go now):
Instead of creating a brand new component `CompactAddToCartComponent` with custom HTML and inherited TS, you could consider reusing the original component and only change it's look-and-feel via pure CSS + pass the different text into it (the icon instead of the default "add to cart" text) via a new `@Input` or preferrably the `<ng-content>` slot.
This would require the following steps:
- add a support of passing a custom content into the original add-to-cart component (via a new `@input` or `<ng-content>` slot)
- reference `<cx-add-to-cart>` instead of `<cx-epd-visualization-compact-add-to-cart>` inside the html template of the parent component `<cx-epd-visualization-compact-add-to-cart>`
- pass into the `<cx-add-to-cart>` the custom content (icon)
- then you could also drop your custom component `cx-epd-visualization-compact-add-to-cart`
- import `AddToCartModule` in `VisualPickingProductListModule` to make the component visible `<cx-add-to-cart>` in the template of the parent component `<cx-epd-visualization-compact-add-to-cart>`

That said, I don't like to go this route in 5.0 due to 2 reasons:
1. in `release/4.3.x` you already merged a version that uses a custom Compact add to cart component. I don't want to introduce inconsistencies, at least in this PR.
2. importing the original `AddToCartModule` inside your  module would unnecessarily provide the cms mapping config for `ProductAddToCartComponent` cms component inside the child injector of your lazy-loaded module. See https://github.com/SAP/spartacus/blob/34c9a5a4d78f76de931e89ef7c305d3b93c2fcc5/feature-libs/cart/base/components/add-to-cart/add-to-cart.module.ts#L11-L20 . Although it wouldn't cause any errors, I believe, this would be semantically incorrect, as only one lazy-loaded chunk should contain a cms mapping for a certain component type. 

    Btw. Now I realized there is an important problem with too many responsibilities of `AddToCartModule`: exposing a component's selector for static usage + providing a cms mapping for it. As a possible improvement in the future we could move out the cms mapping's providers to a separate method so importing the module won't provide the mapping by default. See the example below:
   ```ts
    @NgModule({
      imports: [CommonModule, ReactiveFormsModule, I18nModule, ItemCounterModule],
      declarations: [AddToCartComponent],
      exports: [AddToCartComponent],
    })
    export class AddToCartModule {
      public static forRoot() {
        return {
          ngModule: AddToCartModule,
          providers: [
            provideDefaultConfig(<CmsConfig>{
              cmsComponents: {
                ProductAddToCartComponent: {
                  component: AddToCartComponent,
                  data: {
                    inventoryDisplay: false,
                  },
                },
              },
            }),
          ],
        };
      }
    }
  ```

sam-garland committed:
chore: remove unecessary code. (#15017)

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization

sam-garland commented:
@Platonn - 

I have removed the redundant module references as requested since these are not needed at all.

The issue I was trying to solve with the feature check is not related to importing the `AddToCartComponent` code. The issue is that if the end user runs the schematics in `@spartacus/schematics` and enables the EPD Visualization Integration feature, but does not enable the Cart feature then the `ActiveCartService` implementation is not loaded in the storefront application (since `facadeProviders` is not being included in the `providers` array of any loaded `NgModule`). This leads to visible `CompactAddToCartComponent` buttons that do not do anything when clicked being rendered.

I think it is entirely reasonable to use the visual spare part picking without having the Cart feature enabled (there are no doubt other kinds of actions that might be useful aside from Add to Cart). So I don't think it would be good to make the Cart feature a required dependency of the EPD Visualization Integration feature.

When the Cart feature is not enabled, the customer would most likely need to replace the `VisualPickingProductListComponent` to change the templates. However I was trying to achieve a nicer "out of the box" behaviour when the customer creates a storefront using schematics. I was hoping to have consistent/reasonable behaviour regardless of which other features the user enables.

Using a slot for the button in the list item template would add some additional flexibility, but I would think the same problem would exist when the Cart feature is not enabled. If the slot contains a `CompactAddToCartComponent` button (as it would do by default) the `CompactAddToCartComponent` button would still be rendered but not do anything when clicked.

The change I have made to hide the buttons when the Cart feature is not enabled is not an ideal solution on its own since the list layout still has space reserved for a column that does not contain anything:
![image](https://user-images.githubusercontent.com/52560012/153980627-3faaa808-221d-42e1-9f1d-dfe447fa0e97.png)

Some more special case code would be needed to avoid the ugliness of an extra column. 

Just removing all the special case code would certainly be the simplest option.

Let me know what approach you think is best.

Platonn commented:
@sam-garland 

Thanks for your explanation. Now I understand your motivation for adding the check whether the storefront is capable of having a cart and adding items to it.

TBH I'm not sure the OOTB Spartacus is prepared at this moment for a partial usage of the storefront (without cart functionality). I would need to consult again with @plabadie and @WeizhengSap. Despite the customer can uncheck a  checkbox "Cart" in the CLI while running `ng add @spartacus/schematics`, I don't know whether the rest of OOTB Spartacus will work correctly then. If it's possible, then **we need to find a proper way of checking the existence of the cart functionality.**


I guess the name of the service method is a bit confusing `FeatureModulesService.isConfigured`. Maybe we should  rename to `.isLazyLoadingConfigured` or something similar? 🤷 I'll discuss with the Spartacus core team.


Checking the `featureModulesService.isConfigured(ADD_TO_CART_FEATURE)` means nothing more that the "Add to cart" button component is technically extracted to a tiny dedicated lazy loaded module (apart from the rest of the core cart functionality module) . But the absence of this configuration for the key `ADD_TO_CART_FEATURE` doesn't mean that the cart functionality is not used. The JS for add-to-cart functionality can be theoretically  bundled in a different lazy-loaded module (according to any customizations the customers may have). 

I confirm, this extra configuration for lazy-loading ADD_TO_CART_FEATURE will be created by default by schematics only in the new fresh installed Spartacus apps since 5.0. But for customers migrating their apps manually from Spartacus 4.x to 5.0 this configuration might be missing (unless we document in the migration docs it needs to be added manually or we might create migration-schematics for adding it - that's possible).

Nevertheless, the absence of this config for key ADD_TO_CART_FEATURE doesn't mean semantically that the storefront has disabled the add-to-cart functionality. Therefore I suggest no to rely on this check in your code.

Even if we would check if the `.isConfigured(CART_BASE_FEATURE)` (string key for the heart of the cart mechanism) , we would only check if the base cart module is configured for lazy loading. But not every customer uses lazy loading. I imagine there might be some customers (e.g. migrating from old  Spartacus 1.x or 2.x, where lazy loading was not yet supported nor configurable) that don't configure the lazy loading themselves so quickly after bumping the `@spartacus/*` packages version to 4.x.

I guess you only wanted to make the life of customers easier when they extend your feature and don't use cart functionality, by taking advantage of an existing check `.isConfigured()`, which seemed to be a low hanging fruit. But this condition in the code IMHO doesn't check what you expected.
**Btw. As far as I understand, the support of absent cart functionality wasn't implemented in your feature released in 4.3.0, right?**



I would need to think more and discuss with the Spartacus core team if there exists a proper check to determine the absence of the cart functionality, that you could use in your logic. But I don't want to delay the merge of this PR because of it.

**What do you think about removing the check in this PR? And we could possibly get back to the topic, when we have some idea how to check the existence of the cart functionality correctly?**

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization

sam-garland committed:
Revert "Hide cx-epd-visualization-compact-add-to-cart when addToCart feature not configured (#14973)" (#15044)

This reverts commit bc49b9dcaf38db53914a67f0088825ccebbb1d30.

sam-garland commented:
I have reverted the changes to try to detect whether the Cart feature is enabled given that the check was not doing what I had assumed and cannot be relied upon.

@Platonn - Please let me know if/when there is a better solution for detecting whether the ActiveCartService can be used.

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization

Platonn committed:
Merge branch 'develop' into integration-libs/epd-visualization

Platonn commented on projects/storefrontapp-e2e-cypress/cypress/integration/vendor/cds/profile-tag/default-events.spec.ts:
```diff
@@ -0,0 +1,489 @@
+} from '../../../../helpers/vendor/cds/cds';
+import { profileTagHelper } from '../../../../helpers/vendor/cds/profile-tag';
+
+describe('Profile-tag events', () => {
```
Are those changes really related to this PR?

Platonn replied:
I can see in the PR to 4.3.x we have only changed the path of this file
https://github.com/SAP/spartacus/pull/14753/files#diff-f0e7debfd7f5766491e0bcf6e6e36a7ebaf209e2cc19a22c8002657ab9d2b4c1
<img width="1576" alt="image" src="https://user-images.githubusercontent.com/4001059/154490477-730b9b86-4b2c-418f-9545-f321aaff8cc1.png">

sam-garland replied:
@Platonn

What happened here was that the name of the file was changed in the `develop` branch and in my set of changes and the merge process added the file that had been renamed in my set of changes:

![image](https://user-images.githubusercontent.com/52560012/154574254-5f0fd897-6092-4259-b096-bc8190940c51.png)

I have removed the file added as a result of my renaming so that there is no change introduced relative to the `develop` branch.

Platonn commented on projects/storefrontapp-e2e-cypress/cypress/integration/vendor/epd-visualization/accessibility/visual-picking-tab-tabbing-order.e2e-spec.ts:
```diff
@@ -0,0 +1,14 @@
+  verifyTabbingOrder,
+} from '../../../../helpers/vendor/epd-visualization/visual-picking-tab';
+
+context("Tabbing order - tests don't require user to be logged in", () => {
```
Are those changes really related to this PR?

Platonn replied:
My bad. Looks good.

Platonn commented on projects/storefrontapp-e2e-cypress/cypress/helpers/vendor/epd-visualization/visual-picking-tab.ts:
```diff
@@ -0,0 +1,150 @@
+export function configureDefaultProduct() {
```
Are those changes really related to this PR?

Platonn replied:
My bad. Looks good

Platonn commented on projects/storefrontapp-e2e-cypress/cypress/support/a11y-tab.commands.ts:
```diff
@@ -33,7 +39,7 @@ declare global {
 }
 
-Cypress.Commands.add('pressTab', () => {
+Cypress.Commands.add('pressTab', (shiftModifier?: boolean) => {
```
Are those changes really related to this PR?

Platonn replied:
My bad. Looks good.

Sorry for spam

Platonn commented on projects/storefrontapp-e2e-cypress/cypress/support/utils/a11y-tab.ts:
```diff
@@ -18,3 +18,12 @@ export function getNextFocusableElement(
     : elements[0];
 }
+
+export function getPreviousFocusableElement(
```
Are those changes really related to this PR?

Platonn replied:
My bad. Looks good

Platonn commented:
Hi @sam-garland 
I did the last check to compare changes in PR to 4.x vs. changes in this PR to develop.

I'm sorry I raised a few questions "are those changes really related to this PR" and only later I realized those changes are ok. :( 

There is one change related to CDS' e2e which I believe shouldn't be part of this PR.

Besides that, looks good! Thanks for your updates.

When changes in  CDS' e2e file are sorted out, I'm happy to approve the PR.

sam-garland committed:
chore: fix merge issue. (#15080)

File renamed in `develop` branch.

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization

Platonn committed:
Merge branch 'develop' into integration-libs/epd-visualization

sam-garland committed:
Merge branch 'develop' into integration-libs/epd-visualization
